Sema: Lift flag requirement for access levels on imports now that SE-0409 has been accepted

xymus opened the pull request:
The proposal [SE-0409 Access level on imports](https://github.com/apple/swift-evolution/blob/main/proposals/0409-access-level-on-imports.md) has been accepted, let's lift the flag requirement!

Also adds the upcoming feature flag `InternalImportsByDefault` for imports to default to internal. This is the Swift 6 behavior that can be enabled in Swift 5 using this new flag. The logic is cleaned but by moving all the Swift 6 specific logic being this flag.

rdar://107566870

xymus committed:
[Sema] Intro flag to default imports to internal, the Swift 6 mode

The feature InternalImportsByDefault makes imports default to internal instead
of public. Applying the Swift 6 behavior of SE-0409 in Swift 5.

Let's use only that flag to track the Swift 6 behavior as well instead
of separately checking for the language version.

xymus committed:
[Sema] Update diagnostic about public imports of non-resilient modules

The compiler reports public imports of non-resilient modules from a
resilient module. Make sure that when imports default to internal, which
also implies Swift 6, this is treated as an error and the fixit simply
deletes the `public` keyword.

We keep using the AccessLevelOnImport flag only to report these as
errors in Swift 5 for early adopters.

xymus committed:
[Sema] Lift flag requirement for access level on imports

xymus committed:
[Tests] Remove the flag AccessLevelOnImport from all tests but one

xymus commented:
@swift-ci Please smoke test

tshortli commented on test/ModuleInterface/access-level-import-swiftinterfaces.swift:
```diff
@@ -21,30 +21,57 @@
+/// Swift 6 mode.
+// RUN: %target-swift-frontend -typecheck %t/Client.swift -I %t \
+// RUN:   -package-name TestPackage -module-name Client_Swift6 \
+// RUN:   -enable-library-evolution -swift-version 6 \
```
Don't forget `REQUIRES: asserts`.

xymus replied:
Good catch, thanks!

tshortli commented on include/swift/Basic/Features.def:
```diff
@@ -117,6 +117,7 @@ UPCOMING_FEATURE(BareSlashRegexLiterals, 354, 6)
 UPCOMING_FEATURE(ExistentialAny, 335, 6)
 UPCOMING_FEATURE(ImportObjcForwardDeclarations, 384, 6)
 UPCOMING_FEATURE(DisableOutwardActorInference, 401, 6)
+UPCOMING_FEATURE(InternalImports, 409, 6)
```
This feature name makes it sound like you need to specify it in order to write `internal import`. What about something like `InternalImportsByDefault`?

xymus replied:
Good point, I've updated the PR to use `InternalImportsByDefault`.

tshortli reviewed (approved):
Exciting!

tshortli commented:
This is just a nice to have, but it could be helpful to emit a warning for projects that already adopted `-enable-experimental-feature AccessLevelOnImports` to let them know they can remove the flag.

xymus commented:
I think we'll get the warning automatically when we delete the flag. I'm holding off on that because the flag still has one effect, promoting to errors imports of a non-resilient module from a resilient module. It's a small thing but might as well keep it around until most projects move to Swift 6 mode.

elsh reviewed (approved):
🎆

xymus committed:
[Test] Add public to imports in Swift 6 tests

xymus commented:
@swift-ci Please smoke test

xymus committed:
[Tests] Update swiftinterface imports Swift 6 test

Update the test for imports defaulting to internal by adding the
explicit `public` modifier on all imports and the expected warnings.
Modernized a test somehow as it can't use previously shared files.

This revealed an issue with the warning about superfluous imports on
imports of submodules. I've marked one of them as a FIXME as it will
require a fix in the diagnostic.

xymus commented:
@swift-ci Please smoke test
